pkg/cloud/services/compute/instance.go

```diff
@@ -376,40 +409,71 @@ func getServerNetworks(networkClient *gophercloud.ServiceClient, networkParams [
 	return nets, nil
 }
 
-func getOrCreatePort(is *Service, eventObject runtime.Object, clusterName string, portName string, net infrav1.Network, securityGroups *[]string) (*ports.Port, error) {
+func getOrCreatePort(is *Service, eventObject runtime.Object, clusterName string, portName string, net infrav1.Network, instanceSecurityGroups *[]string) (*ports.Port, error) {
 	allPages, err := ports.List(is.networkClient, ports.ListOpts{
 		Name:      portName,
 		NetworkID: net.ID,
 	}).AllPages()
 	if err != nil {
 		return nil, fmt.Errorf("searching for existing port for server: %v", err)
 	}
-	portList, err := ports.ExtractPorts(allPages)
+	existingPorts, err := ports.ExtractPorts(allPages)
 	if err != nil {
 		return nil, fmt.Errorf("searching for existing port for server: %v", err)
 	}
+	if len(existingPorts) == 0 {
+		description := net.PortOpts.Description
+		if description == "" {
+			description = fmt.Sprintf("Created by cluster-api-provider-openstack cluster %s", clusterName)
+		}
 
-	if len(portList) != 0 {
-		return &portList[0], nil
-	}
+		// inherit port security groups from the instance if not explicitly specified
+		securityGroups := net.PortOpts.SecurityGroups
+		if securityGroups == nil {
+			securityGroups = instanceSecurityGroups
+		}
```

Comment:
Can this logic move to before 126? it is easier to understand constructing security groups(from line 114-126) then construct nets(from line 128-164).

Reply:
The security groups can be set on a port and can be different per port, so I don't think we can put this before 126, before we've looped through the nets and ports in the spec.

SecurityGroups could be added as a field to the Network struct and we could set this default in lines 128-164, instead of in `getOrCreatePort`. I wonder also if constructing nets should be pulled into a separate function, as I agree it takes some understanding. Let me make those changes and you can tell me what you think.

Reply:
Thanks, I understand we can not move this before 126. Please make a separate function as you think.